src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const resultStates = [];
+
+  let stateCopy = Object.assign({}, state);
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties': {
+        Object.assign(stateCopy, action.extraData);
+        break;
+      }
+
+      case 'removeProperties': {
+        for (const key of action.keysToRemove) {
+          delete stateCopy[key];
+        }
+        break;
+      }
+
+      case 'clear': {
+        stateCopy = {};
+        break;
+      }
+
+      default:
+        continue;
+    }
+    resultStates.push({ ...stateCopy });
```

Comment:
It is good that you're creating a new object copy before pushing it to the resultStates array. This ensures that each state in the history is a separate object and not a reference to the same object.